File: commons/tornium_commons/models/torn_key.py
# Copyright (C) 2021-2023 tiksan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import typing

from peewee import (
    BooleanField,
    CharField,
    DeferredForeignKey,
    SmallIntegerField,
    UUIDField,
    fn,
)

from .base_model import BaseModel


class TornKey(BaseModel):
    guid = UUIDField(primary_key=True)
    api_key = CharField(max_length=16, null=False, unique=True, index=True)
    user = DeferredForeignKey("User")
    default = BooleanField(default=False)
    disabled = BooleanField(default=False)
    paused = BooleanField(default=False)  # Comes from Torn API error code 18
    access_level = SmallIntegerField(null=True)

    @classmethod
    def random_key(cls) -> typing.Optional["TornKey"]:
        return (
            TornKey.select(TornKey.api_key)
            .where((TornKey.default == True) & (TornKey.disabled == False) & (TornKey.paused == False))
            .order_by(fn.Random())
            .first()
        )

File: worker/lib/user/schema/torn_key.ex
# Copyright (C) 2021-2025 tiksan
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

defmodule Tornium.Schema.TornKey do
  use Ecto.Schema

  @type access_levels() :: :public | :minimal | :limited | :full

  @type t :: %__MODULE__{
          guid: Ecto.UUID.t(),
          api_key: String.t(),
          user_id: pos_integer(),
          user: Tornium.Schema.User.t(),
          default: boolean(),
          disabled: boolean(),
          paused: boolean(),
          access_level: access_levels()
        }

  @primary_key {:guid, Ecto.UUID, autogenerate: true}
  schema "tornkey" do
    field(:api_key, :string)
    belongs_to(:user, Tornium.Schema.User, references: :tid)
    field(:default, :boolean)
    field(:disabled, :boolean)
    field(:paused, :boolean)
    field(:access_level, Ecto.Enum, values: [public: 1, minimal: 2, limited: 3, full: 4])
  end
end
